### DISCUSSION TITLE
Sign-in with Ethereum (SIWE) support for supabase auth

### DISCUSSION THREAD

**Discussion by Silur** [supabase/supabase#12052] [Feature Requests]
SIWE is an authentication method, very similar (in threat modelling terms) to WebAuthN that has been adopted by auth0 - https://auth0.com/blog/sign-in-with-ethereum-siwe-now-available-on-auth0/ and is used in the web3 scene to extend dApp functionality.

It'd be nice to support SIWE as an auth provider for supabase. I'm not fluent enough in go to handle the nonce generation part in the gotrue repo so I just leave a guide here on how authentication looks like on frontend and backend for SIWE:

https://wagmi.sh/examples/sign-in-with-ethereum

Basically the backend generates a cryptographically secure nonce, which the client signs as:

```javascript
const nonce = await supabase.auth.getSiweNonce()
const message = new SiweMessage({
        domain: window.location.host,
        address,
        statement: 'Sign in with Ethereum to the app.',
        uri: window.location.origin,
        version: '1',
        chainId,
        nonce: state.nonce,
      })
const signature = await signMessageAsync({
        message: message.prepareMessage(),
 })
 
 const { data, error } = await supabase.auth.signInWithEthereum({signature})
```

Once the backend verifies the signature - (that it belongs to its domain, nonce and timestamp is valid etc) it can map a session token to the user wallet that signed the message:

```javascript
import { SiweMessage } from 'siwe'

const handler = async (request, response) => {
       try {
          const { message, signature } = request.body
          const siweMessage = new SiweMessage(message)
          const fields = await siweMessage.validate(signature)
 
          if (fields.nonce !== req.session.nonce)
            return res.status(422).json({ message: 'Invalid nonce.' })
 
          req.session.siwe = fields
          await req.session.save() // assign cookie
          res.json({ ok: true })
       } catch (_error) {
          res.json({ ok: false })
      }
  }

**Comment by devstein** [comment]
Hi @Silur while there is no native support for SIWE from Supabase, there is an integration with [Picket](https://picketapi.com/) to use wallet login with your Supabase project. You can read the integration guide here: https://supabase.com/docs/guides/integrations/picket

**Comment by Silur** [comment]
Thanks, great to know there's a thirdparty provider for this! (could not find picket in the docs by navigating)
Though having a thirdpary, registration only API manage the _decentralized_ login flow for a selfhosted backend kind of kills the whole point of SIWE :sweat_smile:

**Reply by devstein** [reply]
@Silur Not necessarily. Like any managed provider, it's trading convenience for centralization. Picket helps with the onboarding user UX and simplifies backend development, but there is no vendor lock-in. Users own their keys and you can easily migrate to implementing SIWE yourself once at it makes sense for the project.
